casper/src/main/scala/coop/rchain/casper/storage/LmdbStoreManager.scala

```diff
@@ -0,0 +1,138 @@
+      dbs: Map[String, Deferred[F, DbEnv[F]]] = Map.empty
+  ) {
+    override def toString() = s"DbState(status: $status, inProgress: $inProgress)"
+  }
+
+  // Internal manager state for LMDB databases and environments
+  private val varState = Ref.unsafe(DbState(EnvClosed, 0, envDefer))
+
+  override def database(dbName: String): F[KeyValueStore[F]] =
+    Log[F].info(s"LMDB database created path: $dirPath, DB: $dbName").pure[F] >>
+      Sync[F].delay(new LmdbKeyValueStore[F](getCurrentEnv(dbName)))
+
+  private def getCurrentEnv(dbName: String): F[DbEnv[F]] =
+    for {
+      // Check if environment is closed
+      isNewEnv <- varState.modify { st =>
+                   if (st.status == EnvClosed) {
+                     val newState = st.copy(status = EnvStarting)
+                     (newState, true)
+                   } else (st, false)
+                 }
+      // Create new environment when it's closed.
+      // IMPORTANT: this can happen only once, all other callers
+      // are blocked on Deferred object until it's completed.
+      _ <- createEnv.whenA(isNewEnv)
+
+      // Block until environment is ready.
+      st  <- varState.get
+      env <- st.envDefer.get
+
+      // Find DB ref / first time create deferred object.
+      newDbDefer <- Deferred[F, DbEnv[F]]
+      action <- varState.modify { st =>
+                 val maybeDbRef = st.dbs.get(dbName)
+                 maybeDbRef.fold {
+                   val newDbs = st.dbs.updated(dbName, newDbDefer)
+                   val newSt  = st.copy(dbs = newDbs)
+                   (newSt, (true, newDbDefer))
+                 }(defer => (st, (false, defer)))
+               }
+
+      // If database is not found, create new Deferred object to block
+      // callers until database is created and Deferred object completed.
+      (isNewDb, dbEnvDefer) = action
+      _ <- (for {
+            dbi <- Sync[F].delay(env.openDbi(dbName, DbiFlags.MDB_CREATE))
+            _   <- dbEnvDefer.complete(DbEnv(env, dbi, decrementDbRefCounter))
+          } yield ()).whenA(isNewDb)
+
+      // Block until database is ready.
+      envDbi <- dbEnvDefer.get
+
+      // Increment request counters to track in unfinished requests.
+      _ <- incrementDbRefCounter
+    } yield envDbi
+
+  // Begin database request
+  private def incrementDbRefCounter: F[Unit] =
+    varState.update { st =>
+      st.copy(inProgress = st.inProgress + 1)
+    }
+
+  // Finish database request
+  private def decrementDbRefCounter: F[Unit] =
+    varState.update { st =>
+      st.copy(inProgress = st.inProgress - 1)
+    }
+
+  // Create LMDB environment and update the state
+  private def createEnv: F[Unit] =
+    for {
+      _     <- Sync[F].delay(Files.createDirectories(dirPath))
+      flags = Seq(EnvFlags.MDB_NOTLS)
+      // Create environment
+      env <- Sync[F].delay(
+              Env
+                .create()
+                .setMapSize(10L * 1024L * 1024L * 1024L)
```

Comment:
@zsluedem the idea is for RNodeKeyValueManager to accepts configuration for all stores and instantiate LmdbStoreManager parameters like map size or database count. We can do this as part of PR where it will be used first.